### PR TITLE
Expose `SpkIterator::new_with_range`

### PR DESCRIPTION
**Pull request by evanlinjin**
### Description

This method is useful, I'm not sure why it is not public.

I've also updated the docs and tests.

### Changelog notice

Added
* Expose `SpkIterator::new_with_range`

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/chain/src/spk_iter.rs [hunk 1]
```diff
@@ -43,18 +43,24 @@ impl<D> SpkIterator<D>
...
-    // Creates a new script pubkey iterator from a descriptor with a given range.
-    // If the descriptor doesn't have a wildcard, we shorten whichever range you pass in
-    // to have length <= 1. This means that if you pass in 0..0 or 0..1 the range will
-    // remain the same, but if you pass in 0..10, we'll shorten it to 0..1
-    // Also note that if the descriptor doesn't have a wildcard, passing in a range starting
-    // from n > 0, will return an empty iterator.
-    pub(crate) fn new_with_range<R>(descriptor: D, range: R) -> Self
+    /// Create a new script pubkey iterator from `descriptor` and a given `range`.
+    ///
+    /// The end bound of the range will be shortened to 0 (inclusive) for non-wildcard descriptors
```
Comment (LLFourn) [inline crates/chain/src/spk_iter.rs]: "the end bound of the range will be shortened to 0 (inclusive)" is a very complicated way of saying "non-wildcard descriptors will only emit a single item".

File: crates/chain/src/spk_iter.rs [hunk 2]
```diff
@@ -73,11 +79,9 @@ where
...
         end = end.min(BIP32_MAX_INDEX + 1);
 
-        if !descriptor.borrow().has_wildcard() {
-            // The length of the range should be at most 1
-            if end != start {
-                end = start + 1;
-            }
+        // A non-wildcard descriptor will only have index 0.
+        if !descriptor.borrow().has_wildcard() && end > 1 {
+            end = 1;
```
Comment (LLFourn) [inline crates/chain/src/spk_iter.rs]: what happens if we just remove this. It seems to be checked in `next`. Looks unnecessary.

**Review by LLFourn — CHANGES_REQUESTED** [review]

File: crates/chain/src/spk_iter.rs (hunk 2 above)
Comment (evanlinjin) [reply crates/chain/src/spk_iter.rs]: Removed.

File: crates/chain/src/spk_iter.rs [hunk 3]
```diff
@@ -250,6 +249,14 @@ mod test {
             SpkIterator::new_with_range(&no_wildcard_descriptor, 1..=2).next(),
             None
         );
+        assert_eq!(
```
Comment (danielabrozzoni) [inline crates/chain/src/spk_iter.rs:252]: I'm not sure why these tests are being added, but I won't complain about adding new tests 😛

**Review by danielabrozzoni — APPROVED** [review]
utACK 8305e648494cbdd1f6438904ec367f8442fba606

**Review by LLFourn — APPROVED** [review]
ACK 8305e648494cbdd1f6438904ec367f8442fba606

**Review by LLFourn — APPROVED** [review]
DELETED COMMENT

**Pull request merged**